### DISCUSSION TITLE
[Best practice] balance between number of packages and performance

### DISCUSSION THREAD

**Discussion by leonard-henriquez** [vercel/turborepo#6347] [Help]
My team is currently facing a dilemma regarding the balance we should strike between performance and the number of packages.

We currently have approximately 20 packages for a medium-sized codebase (4 engineers working full time over the last year).
Our goal is to maintain a modular codebase with packages that are as small and manageable as possible.
However, we have concerns about the potential impact on performance when running too many packages concurrently, particularly when executing the `build --watch` command for all packages in development mode.
We've noted that running our `turbo run dev` command already consumes a significant amount of resources on recent MacBooks.

Looking ahead, we are questioning the scalability of our approach. 
Would it be feasible to manage 100+ packages?
Would we still be able to execute the `turbo run dev` command locally without running into performance bottlenecks?
We're curious if you have any recommendations or best practices to share.

**Comment by anthonyshew** [comment]
A few initial thoughts:
- You can ensure you're only running the development tasks that you need by using `--filter` to only run the tasks that are in the graph of the application that you are working. ([Documentation](https://turbo.build/repo/docs/reference/command-line-reference/run#--filter))
- It's possible you may not need `dev` tasks for your package dependencies at all if you are using modern frameworks. Using the Internal Package pattern, you can let your framework handle the work of compiling/bundling your packages themselves. ([Documentation](https://turbo.build/repo/docs/handbook/sharing-code/internal-packages))

**Reply by leonard-henriquez** [reply]
Thank you for your answer but it doesnt really answer my question:
- I know about filters but I don't want to use them too much in dev mode because I don't want my developer to need to restart their turbo every time they need to modify a file in another package -> my question is actually about wether I can run 100 packages at the same time, not how to find a way around it 
- Bundling through the end application is not an option in my case (and I guess it's not for most of the complex monorepos where you have multiple apps consuming the same shared packages)

**Reply by anthonyshew** [reply]
As long as your machine can handle the workload you're asking for, then I suppose you could run as many dev tasks as needed. Tasks vary on how demanding they are.

 It would depend on the specifications of the machine and how heavy the tasks you're asking to run are. `turbo` itself will be only a small sliver of the work. The actual processes will be the bulk.

**Reply by leonard-henriquez** [reply]
Ok thank you for your answer.
However it still doesn't really answer my question.
I'm not sure it's only related to the specification of my machine.
The number of packages that you should have greatly depends on the fixed cost associated with each extra task/package.
As most dev machines these days have from 8 to 32 GB or RAM it give a pretty good order of magnitude of the number of packages that you can handle.

So I can understand that you don't have the numbers, but it's a bummer that turbo doesn't provide benchmarks on this topic.
When I choose a tool manage my monorepo I definitely expect this tool to have an opinion on how to split my packages...

**Reply by anthonyshew** [reply]
We can't have an opinion because the work being performed by a repo is inherently variable.

Not all tasks are created equal, not all workspaces are created equal, not all repos are created equal...There's no one true answer. There's no way for Turborepo to administer your design specifications nor do we dictate the tooling you choose to use in your Turborepo. You may be building a utility-driven package, a domain-driven package, an application package, a library package, a TypeScript package, a JavaScript package...All of these things vary.

**Reply by leonard-henriquez** [reply]
Even if you can't provide one size fit all answer you could provide informations, benchmarks or rule of thumb based on some key parameters (number of cpu?, quantity of RAM?). Having absolutely no information tends to: 
1) complexify the (already difficult) architectural decisions
2) push people to take wrong decisions based on information they don't have

**Reply by anthonyshew** [reply]
I'm not sure what sort of benchmark or information you're suggesting. The overhead of running `turbo` will be insignificant (as in, likely not even 1%) resource usage compared to the tasks themselves.

Given that, it's much more important that you benchmark the resource usage of the frameworks/tooling in your repo. Plenty of tools exist for doing this today so I'm not sure what role `turbo` could serve to help.
